pkg/auth/oauth/registry/timeoutvalidator.go

```diff
@@ -0,0 +1,267 @@
+	tokens         oauthclient.OAuthAccessTokenInterface
+	tokenChannel   chan tokenData
+	data           map[string]tokenData
+	tree           *btree.BTree
+	defaultTimeout time.Duration
+	flushTimeout   time.Duration
+	safetyMargin   time.Duration
+}
+
+func NewTimeoutValidator(tokens oauthclient.OAuthAccessTokenInterface, oauthClient OAuthClientGetter, defaultTimeout int32) *TimeoutValidator {
+	// flushTimeout is set to one third of defaultTimeout
+	flushTimeout := defaultTimeout / 3
+	if flushTimeout < validation.MinFlushTimeout {
+		flushTimeout = validation.MinFlushTimeout
+	}
+	// safetyMargin is set to one tenth of flushTimeout
+	safetyMargin := flushTimeout / 10
+	ttu := &TimeoutValidator{
+		oauthClient,
+		tokens,
+		make(chan tokenData),
+		make(map[string]tokenData),
+		// FIXME: what is the right degree for the btree
+		btree.New(32),
+		timeoutAsDuration(defaultTimeout),
+		timeoutAsDuration(flushTimeout),
+		timeoutAsDuration(safetyMargin),
+	}
+	glog.V(5).Infof("Token Timeout Updater primed with defaultTimeout=%d flushTimeout=%d safetyMargin=%d", defaultTimeout, flushTimeout, safetyMargin)
+	return ttu
+}
+
+func timeoutAsDuration(timeout int32) time.Duration {
+	return time.Duration(timeout) * time.Second
+}
+
+// CheckTimeout is called with a token when it is seen by an authenticator
+// it touches only the tokenChannel so it is safe to call from other threads
+func (a *TimeoutValidator) Validate(token *oauth.OAuthAccessToken, _ *user.User) error {
+	if token.TimeoutsIn == 0 {
+		// We care only if the token was created with a timeout to start with
+		return nil
+	}
+
+	td := tokenData{
+		token: token,
+		seen:  time.Now(),
+	}
+	if td.timeout().Before(td.seen) {
+		return ErrTimedout
+	}
+
+	// After a positive timeout check we need to update the timeout and
+	// schedule an update so that we can either set or update the Timeout
+	// we do that launching a micro goroutine to avoid blocking
+	go a.updateToken(td)
+
+	return nil
+}
+
+func (a *TimeoutValidator) updateToken(td tokenData) {
+	a.tokenChannel <- td
+}
+
+func (a *TimeoutValidator) updateTimeouts(oauthClient *oauth.OAuthClient) {
+	timeout := *oauthClient.AccessTokenTimeoutSeconds / 3
+	flushTimeout := int32(a.flushTimeout / time.Second)
+	if timeout < flushTimeout {
+		if timeout < validation.MinFlushTimeout {
+			timeout = validation.MinFlushTimeout
+		}
+		a.flushTimeout = timeoutAsDuration(timeout)
+		a.safetyMargin = timeoutAsDuration(timeout / 10)
+		glog.V(5).Infof("OauthClient %q updates timeouts to flushTimeout=%s safetyMargin=%s", oauthClient.Name, a.flushTimeout, a.safetyMargin)
+	}
+}
+
+func (a *TimeoutValidator) clientTimeout(name string) time.Duration {
+	oauthClient, err := a.oauthClient.Get(name)
+	if err != nil {
+		glog.V(5).Infof("Failed to fetch OAuthClient %q for timeout value: %v", name, err)
+		return a.defaultTimeout
+	}
+	if oauthClient.AccessTokenTimeoutSeconds == nil {
+		return a.defaultTimeout
+	}
+	a.updateTimeouts(oauthClient)
```

Comment:
I believe I had some extra glogs in this func and some other places in this file?

Reply:
I did not import all of them as some things simply changed, any specific thing you want to generate logs out of ? You can describe the thought process about what you think should be logged too.